go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -0,0 +1,293 @@
+	"vitess.io/vitess/go/vt/vttablet/tabletconn"
+
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+type vcopier struct {
+	vr        *vreplicator
+	tablePlan *TablePlan
+}
+
+func newVCopier(vr *vreplicator) *vcopier {
+	return &vcopier{
+		vr: vr,
+	}
+}
+
+func (vc *vcopier) initTablesForCopy(ctx context.Context) error {
+	defer vc.vr.dbClient.Rollback()
+
+	plan, err := buildReplicatorPlan(vc.vr.source.Filter, vc.vr.tableKeys, nil)
+	if err != nil {
+		return err
+	}
+	if err := vc.vr.dbClient.Begin(); err != nil {
+		return err
+	}
+	// Insert the table list only if at least one table matches.
+	if len(plan.TargetTables) != 0 {
+		var buf strings.Builder
+		buf.WriteString("insert into _vt.copy_state(vrepl_id, table_name) values ")
+		prefix := ""
+		for name := range plan.TargetTables {
+			fmt.Fprintf(&buf, "%s(%d, %s)", prefix, vc.vr.id, encodeString(name))
+			prefix = ", "
+		}
+		if _, err := vc.vr.dbClient.ExecuteFetch(buf.String(), 1); err != nil {
+			return err
+		}
+		if err := vc.vr.setState(binlogplayer.VReplicationCopying, ""); err != nil {
+			return err
+		}
+	} else {
+		if err := vc.vr.setState(binlogplayer.BlpStopped, "There is nothing to replicate"); err != nil {
+			return err
+		}
+	}
+	return vc.vr.dbClient.Commit()
+}
+
+func (vc *vcopier) copyNext(ctx context.Context, settings binlogplayer.VRSettings) error {
+	qr, err := vc.vr.dbClient.ExecuteFetch(fmt.Sprintf("select table_name, lastpk from _vt.copy_state where vrepl_id=%d", vc.vr.id), 10000)
+	if err != nil {
+		return err
+	}
+	var tableToCopy string
+	copyState := make(map[string]*sqltypes.Result)
+	for _, row := range qr.Rows {
+		tableName := row[0].ToString()
+		lastpk := row[1].ToString()
+		if tableToCopy == "" {
+			tableToCopy = tableName
+		}
+		copyState[tableName] = nil
+		if lastpk != "" {
+			var r querypb.QueryResult
+			if err := proto.UnmarshalText(lastpk, &r); err != nil {
+				return err
+			}
+			copyState[tableName] = sqltypes.Proto3ToResult(&r)
+		}
+	}
+	if len(copyState) == 0 {
+		return fmt.Errorf("unexpected: there are no tables to copy")
+	}
+	if err := vc.catchup(ctx, copyState); err != nil {
+		return err
+	}
+	return vc.copyTable(ctx, tableToCopy, copyState)
+}
+
+func (vc *vcopier) catchup(ctx context.Context, copyState map[string]*sqltypes.Result) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	settings, err := binlogplayer.ReadVRSettings(vc.vr.dbClient, vc.vr.id)
+	if err != nil {
+		return err
+	}
+	if settings.StartPos.IsZero() {
+		return nil
+	}
+
+	// Start vreplication.
+	errch := make(chan error, 1)
+	go func() {
+		errch <- newVPlayer(vc.vr, settings, copyState, mysql.Position{}).play(ctx)
+	}()
+
+	// Wait for catchup.
+	tmr := time.NewTimer(1 * time.Second)
+	seconds := int64(replicaLagTolerance / time.Second)
+	defer tmr.Stop()
+	for {
+		sbm := vc.vr.stats.SecondsBehindMaster.Get()
+		if sbm < seconds {
+			cancel()
+			<-errch
```

Comment:
I'm a bit nervous about whether there could be some error case where SecondsBehindMaster gets set to 0 and vplayer keeps erroring out without updating SecondsBehindMaster, which could appear to this code as if vplayer ran successfully and there's no delay.

The legacy binlog player can do that kind of thing (failing to update seconds behind master when it's in an error state)

Reply:
I was worried about this also, which is why I did this PR first, to track this more accurately: https://github.com/vitessio/vitess/pull/4637.
But it's something we have to keep an eye one. There could be other corner cases.